File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeZeroSumSublists(ListNode* head) {
        if(!head) return NULL;

        int sum = 0;
        ListNode *h = head;
        while(h){
            sum += h->val;

            if(sum == 0){
                head = h->next;
                return removeZeroSumSublists(head);
            }
            h = h->next;
        }

        head->next = removeZeroSumSublists(head->next);
        return head;
    }
};

File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        if(head == null) return null;

        int sum = 0;
        ListNode node = head;
        while(node != null){
            sum += node.val;

            if(sum == 0){
                head = node.next;
                return removeZeroSumSublists(head);
            }
            node = node.next;
        }
        head.next = removeZeroSumSublists(head.next);

        return head;
    }
}